### PR TITLE
Donate Block – amounts and frequencies customisation

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Allows for customisation of donation amounts and frequencies. 

Closes https://github.com/Automattic/newspack-blocks/issues/741 

### How to test the changes in this Pull Request:

_Testing instructions in https://github.com/Automattic/newspack-plugin/pull/1753_

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by thomasguillot** [comment]
I've pushed a few changes in https://github.com/Automattic/newspack-blocks/pull/1191/commits/f80e6775b70f3fa7f8784169486a0d59cd929d0c

Instead of having the input directly in the editor, is there any way we could use conditional `PanelBody` for this?

eg-donations-sidebar

**Comment by adekbadek** [comment]
That's a tradeoff between a cleaner UI and discoverability. Panels are not great for discoverability and users might never know these amounts are customisable in manual mode. On the other hand, these are overrides, not meant to be frequently used. I think putting these in the panels is fine if the "main" panel clearly communicates where the amounts can be customised, if manual mode is selected. WDYT?

**Review by dkoo — CHANGES_REQUESTED** [review]
This is a big refactor and makes a lot of improvements! I also love seeing new TS examples in action that will be valuable for future reference.

I ran into a few odditites while testing—more details below.

File: src/blocks/donate/editor.scss [hunk 1]
```diff
@@ -33,6 +33,28 @@
...
+				position: absolute;
+				max-width: 81px;
+				top: -31px;
+				z-index: 9;
+				left: 5px;
+				display: none;
+			}
+
+			&:checked ~ input[type='number'] {
+				display: flex;
```
Comment (dkoo) [inline src/blocks/donate/editor.scss:52]: Currently, the input of the last-highlighted tier will remain visible if you click outside of the block. If the block isn't currently selected in the editor, it should probably hide all of these inputs.

File: src/blocks/donate/view.scss [hunk 2]
```diff
@@ -461,12 +459,6 @@
...
 	&.untiered {
-		padding-top: calc( 3 * ( 0.76rem + 1.28em + 4px ) );
-
-		@include media( tablet ) {
-			padding-top: calc( 0.76rem + 1.28em + 4px );
-		}
-
 		.input-container {
 			margin: 0.56rem 0 1.12rem;
 			text-align: left;
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply src/blocks/donate/view.scss:459]: This also happened with Minimal style. Fixed in 89afdd6af4fa1e59905ecc9b3f0513e9cd4b5dbe

File: src/blocks/donate/view.scss [hunk 3]
```diff
@@ -291,7 +272,6 @@
 		@include media( tablet ) {
 			border-left-width: 1px;
 			border-right-width: 1px;
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply src/blocks/donate/view.scss]: Good catch! Fixed in 3d847d006871cc820975c595f77c799e0f13dc39

File: src/blocks/donate/edit.tsx [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+				<PanelBody title={ __( 'Suggested Donations', 'newspack-blocks' ) }>
+					<SelectControl
+						label={ __( 'Default Tab', 'newspack' ) }
+						value={ attributes.defaultFrequency }
+						options={ availableFrequencies.map( key => ( {
+							label: FREQUENCIES[ key ],
+							value: key,
+						} ) ) }
+						onChange={ ( defaultFrequency: FrequencySlug ) =>
+							setAttributes( { defaultFrequency } )
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply src/blocks/donate/edit.tsx:452]: The syntax is correct here – it's not object destructuring (`( { var1: _var1 } )`), but a TS type annotation (`( var: VarType )`). 

But you're right that there was a regression here. When the default frequency was updated, the tab in the form was selected. I've added this functionality in be368dd31aef91d59998a7197d41e3da82c4ef8d.

File: src/blocks/donate/editor.scss (hunk 1 above)
Comment (adekbadek) [reply src/blocks/donate/editor.scss:61]: This UI has been changed as per [earlier discussion](https://github.com/Automattic/newspack-blocks/pull/1191#issuecomment-1172430028) on this PR.

**Comment by adekbadek** [comment]
@thomasguillot @dkoo – alternative proposition for the amounts overriding UI:

image

**Comment by adekbadek** [comment]
> Alternatively, we could move the inputs into 2 rows/columns (but I still think it's a bit cramped with a lack of hierarchy)

I like this idea, done in 44053c76266d9d8212701cd4caa93c5a96ce294f. I think it's more intuitive, with the settings in one place, than dividing donation frequencies into panels.

File: src/blocks/donate/edit.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (dkoo) [reply src/blocks/donate/edit.tsx:471]: Thanks for explaining the TS syntax! The regression is still reproducible for me, though. To replicate:

1. On `master`, add a Donate block to a post and leave the default frequency as-is (defaults to "monthly"). Save the post.
2. Switch to this branch and the corresponding branch in Newspack Plugin.
3. Refresh the editor. Observe that the block's default frequency is now "One Time" in the editor, even though it still shows the "Monthly" tab by default on the front-end.

**Review by dkoo — CHANGES_REQUESTED** [review]
Thanks for the changes! Looks like the only remaining issue is the [default frequency regression](https://github.com/Automattic/newspack-blocks/pull/1191#discussion_r914004942).

File: src/blocks/donate/edit.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (adekbadek) [reply src/blocks/donate/edit.tsx:471]: Ah, I misunderstood what you've meant by "always gets reset". I thought it's about the block in the editor switching to the frequency when the default frequency attribute is updated – which was _also_ a regression.

Fixed in 5ed9720fef1fe8ccb9d8c5c0727063a6393df5d7

[the remark this answers is not in the document]
Comment (dkoo) [reply src/blocks/donate/edit.tsx:471]: Ah, I see what you mean. I can confirm that both regressions appear fixed now:

* When I update the default frequency in block attributes, the editor preview selects that frequency
* When migrating blocks from `master` that had the default `monthly` frequency set, they're migrated correctly with the same frequency option selected in the editor.

**Review by dkoo — APPROVED** [review]
Looking good now!

**Pull request merged**